zipline/_protocol.pyx

```diff
@@ -76,6 +79,289 @@ cdef class BarData:
+        """
+        Returns a window of data for the given assets and fields.
+
+        This data is adjusted for splits, dividends, and mergers as of the
+        current algorithm time.
+
+        The semantics of missing data are identical to the ones described in
+        the notes for `get_spot_value`.
+
+        Parameters
+        ----------
+        assets: Asset or iterable of Asset
+
+        fields: string or iterable of string.  Valid values are "open", "high",
+            "low", "close", "volume", "price", and "last_traded".
+
+        bar_count: integer number of bars of trade data
+
+        frequency: string. "1m" for minutely data or "1d" for daily date
+
+        Returns
+        -------
+        series or DataFrame or Panel, depending on the dimensionality of
+            the 'assets' and 'fields' parameters.
+
+            If single asset and field are passed in, the returned Series is
+            indexed by dt.
+
+            If multiple assets and single field are passed in, the returned
+            DataFrame is indexed by dt, and has assets as columns.
+
+            If a single asset and multiple fields are passed in, the returned
+            DataFrame is indexed by dt, and has fields as columns.
+
+            If multiple assets and multiple fields are passed in, the returned
+            Panel is indexed by field, has dt as the major axis, and assets
+            as the minor axis.
+        """
+        if isinstance(fields, str):
+            single_asset = isinstance(assets, Asset)
+
+            if single_asset:
+                asset_list = [assets]
+            else:
+                asset_list = assets
+
+            df = self.data_portal.get_history_window(
+                asset_list,
+                self.simulation_dt_func(),
+                bar_count,
+                frequency,
+                fields
+            )
+
+            if single_asset:
+                # single asset, single field, return a series.
+                return df[assets]
+            else:
+                # multiple assets, single field, return a dataframe whose
+                # columns are the assets, indexed by dt.
+                return df
+        else:
+            if isinstance(assets, Asset):
+                # one asset, multiple fields. for now, just make multiple
+                # history calls, one per field, then stitch together the
+                # results. this can definitely be optimized!
+
+                # returned dataframe whose columns are the fields, indexed by
+                # dt.
+                return pd.DataFrame({
+                    field: self.data_portal.get_history_window(
+                        [assets],
+                        self.simulation_dt_func(),
+                        bar_count,
+                        frequency,
+                        field
+                    )[assets] for field in fields
+                })
+            else:
+                df_dict = {
```

Comment:
Have not profiled, but I the extra DataFrame creation per sid at a large number will be expensive. We should look at either having `get_history_window` return a `Panel` or 3-dim ndarray.